core/src/main/scala/kafka/log/LogManager.scala

```diff
@@ -623,37 +682,108 @@ class LogManager(logDirs: Array[File],
   }
 
   /**
+    * Mark the partition directory in the source log directory for deletion and
+    * rename the future partition in the destination log directory to be the primary partition
+    *
+    * @param topicPartition TopicPartition that needs to be swapped
+    */
+  def deleteAndSwapFutureLog(topicPartition: TopicPartition): Unit = {
+    logCreationOrDeletionLock synchronized {
+      val sourceLog = primaryLogs.get(topicPartition)
+      val destLog = futureLogs.get(topicPartition)
+
+      var sourceRenamedForDelete = true
+      var destRenamedForPrimary = true
+      var errorMsgOpt: Option[String] = None
+      if (sourceLog.renameDir(Log.logDeleteDirName(topicPartition))) {
+        if (!destLog.renameDir(Log.logPrimaryDirName(topicPartition))) {
+          logDirFailureChannel.maybeAddOfflineLogDir(destLog.dir.getParent, s"Failed to rename dir for partition $topicPartition",
+            new IOException(s"Failed to rename dir for partition $topicPartition"))
+          // Revert the source log directory name so that it will not be deleted
+          if (sourceLog.renameDir(Log.logPrimaryDirName(topicPartition))) {
+            sourceRenamedForDelete = false
+            destRenamedForPrimary = false
+            errorMsgOpt = Some(s"Failed to rename destination log directory ${destLog.dir.getAbsolutePath} to be the primary partition")
+          } else {
+            sourceRenamedForDelete = true
+            destRenamedForPrimary = false
+            logDirFailureChannel.maybeAddOfflineLogDir(sourceLog.dir.getParent, s"Failed to rename dir for partition $topicPartition",
+              new IOException(s"Failed to rename dir for partition $topicPartition"))
+            errorMsgOpt = Some(s"Failed to rename destination log directory ${destLog.dir.getAbsolutePath} to " +
+              s"be the primary partition and failed to rename source log directory ${sourceLog.dir.getAbsolutePath} to avoid deletion")
+          }
+        }
+      } else {
+        logDirFailureChannel.maybeAddOfflineLogDir(sourceLog.dir.getParent, s"Failed to rename dir for partition $topicPartition",
+          new IOException(s"Failed to rename dir for partition $topicPartition"))
+        sourceRenamedForDelete = false
+        destRenamedForPrimary = false
+        errorMsgOpt = Some(s"Failed to rename source log directory ${sourceLog.dir.getAbsolutePath} for deletion")
+      }
+
+      /*
+       * It is possible for the log of this partition to be mistakenly deleted from this broker if the following sequence
+       * of events happen. However, the chance of this happening is extremely low. And if the replication factor > 1, the broker
+       * should be able to re-fetch the log for this partition from the leader broker if the log is deleted by mistake.
+       *
```

Comment:
But if replication factor == 1 we've lost the log? From the description you give it sounds like we lose track of the state of the renaming because the only information to go on are the names of the files in the directory. Could this be solved by having a transaction log file for the deletion in the same directory?

Reply:
I will comment on this later.